os/drivers/testcase/test_group.c

```diff
@@ -0,0 +1,263 @@
+	if (child != child_returned) {
+		dbg("group_removechild failed.");
+		goto errout_with_addchild;
+	}
+
+	child_returned = group_findchild(group, child_pid);
+	if (child_returned != NULL) {
+		dbg("group_removechild failed.");
+		group_removechild(group, child_pid);
+		group_freechild(child);
+		return ERROR;
+	}
+	group_freechild(child);
+	return OK;
+
+errout_with_addchild:
+	group_removechild(group, child_pid);
+	group_freechild(child);
+	return ERROR;
+}
+
+static int test_group_alloc_free(int cmd, unsigned long arg)
+{
+	int ret;
+	struct child_status_s child_dummy;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child = group_allocchild();
+	if (child == NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+	if (child->flink != NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+
+	child->flink = &child_dummy;
+	group_freechild(child);
+	if (child->flink == &child_dummy) {
+		dbg("group_freechild failed.");
+		group_freechild(child);
+		return ERROR;
+	}
+	return OK;
+}
+
+static int test_group_exit_child(int cmd, unsigned long arg)
+{
+	int ret;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child_pid = kernel_thread("group", SCHED_PRIORITY_DEFAULT, TASK_STACKSIZE, group_exitchild_func, (char *const *)NULL);
+	if (child_pid < 0) {
+		dbg("task_create failed.");
+		return ERROR;
+	}
+
+	child = group_findchild(group, child_pid);
+	if (child == NULL) {
+		dbg("child is null.");
+		return ERROR;
+	}
+
+	sleep(3);
+
+	child_returned = group_exitchild(group);
+	if (child != child_returned) {
+		dbg("group_exitchild failed.");
+		return ERROR;
+	}
+
+	child_returned = group_removechild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_removechild failed.");
+		return ERROR;
+	}
+
+	group_freechild(child);
+	return OK;
+}
+
+static int test_group_removechildren(int cmd, unsigned long arg)
+{
+	int ret;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child = group_allocchild();
+	if (child == NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+
+	child_pid = -1;
+	child->ch_flags = TCB_FLAG_TTYPE_TASK;
+	child->ch_pid = child_pid;
+	child->ch_status = 0;
+	/* Add the entry into the TCB list of children */
+	group_addchild(group, child);
+
+	/* cross check starts */
+	child_returned = group_findchild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_findchild failed.");
+		goto errout_with_addchild;
+	}
+
+	group_removechildren(group);
+	if (group->tg_children != NULL) {
+		dbg("group_removechildren failed.");
+		goto errout_with_addchild;
+	}
+	group_freechild(child);
+	return OK;
+
+errout_with_addchild:
+	group_removechildren(group);
+	group_freechild(child);
+	return ERROR;
+}
+
+static int test_init(void)
+{
+	st_tcb = sched_self();
+	if (st_tcb == NULL) {
+		dbg("sched_self failed.");
+		return ERROR;
+	}
+
+	group = st_tcb->group;
+	if (group == NULL) {
+		dbg("group is null.");
```

Comment:
dbg is for general module so please change XXdbg for this module instead of dbg.